lib/net/http/header.rb

```diff
@@ -1,16 +1,159 @@
 # frozen_string_literal: false
-# The HTTPHeader module defines methods for reading and writing
-# HTTP headers.
 #
-# It is used as a mixin by other classes, to provide hash-like
-# access to HTTP header values. Unlike raw hash access, HTTPHeader
-# provides access via case-insensitive keys. It also provides
-# methods for accessing commonly-used HTTP header values in more
-# convenient formats.
+# The \HTTPHeader module provides access to \HTTP headers.
+# The headers are a hash-like collection of key/value pairs called _fields_.
+#
+# The module is included in:
+#
+# - Net::HTTPGenericRequest (and therefore Net::HTTPRequest).
+# - Net::HTTPResponse.
+#
+# == About the Examples
+#
+# :include: doc/net-http/examples.rdoc
+#
+# == Fields
+#
+# A header field is a key/value pair.
+#
+# === Field Keys
+#
+# A field key may be:
+#
+# - A string: Key <tt>'Foo'</tt> is treated as if it were
+#   <tt>'Foo'.downcase</tt>;  i.e., <tt>'foo'</tt>.
+# - A symbol: Key <tt>:Foo</tt> is treated as if it were
+#   <tt>:Foo.to_s.downcase</tt>;  i.e., <tt>'foo'</tt>.
+#
+# Examples:
+#
+#   req = Net::HTTP.get(uri)
+#   req[:Accept]  # => "*/*"
+#   req['Accept'] # => "*/*"
+#   req['ACCEPT'] # => "*/*"
+#
+#   req['accept'] = 'text/html'
+#   req['accept'] # => "text/html"
+#   req[:Accept] = 'text/html'
+#   req['accept'] # => "text/html"
+#   req['Accept'] = 'application/json'
+#   req['accept'] # => "application/json"
+#   req['ACCEPT'] = 'text/plain'
+#   req['accept'] # => "text/plain"
+#
+# === Field Values
```

Comment:
So, no change requested here, right?

Reply:
Well this section implicitly shows how arrays and hashes are converted to strings, maybe we can move that to a separate section to highlight that behavior?

Reply:
I don't know what you mean.   The example field values has give segments.  That's not explicit enough?

Reply:
What I mean is that it was a surprise for me that we can have values of arrays and hashes since HTTP headers don't allow these types (it's a non-standard thing Ruby is doing). How it converts the arrays and hashes to string is also surprising to me (even more so for hashes, since it creates an array of keys and values). I think it's worth highlighting since IMO it's very surprising behavior.

Reply:
I was surprised, too.  But I think that the separate examples for arrays, hashes, and mixtures is doing that work.  You may have to show me how you think this should be.

Reply:
If you think the explanation is sufficient, then I'm ok with it 👍